### ISSUE TITLE
Evaluate Octen-Embedding-8B as retrieval model upgrade

### ISSUE THREAD

**Issue by EDI-moltbot** [pythagorakase/nexus#175]
## Context

Current NEXUS retrieval pipeline uses `inf-retriever-v1-1.5b` (1.5B params, 1536 dims).

Recent RTEB (Retrieval Text Embedding Benchmark) results show significant improvements in newer models, particularly in the Octen family:

| Model | Params | Dims | RTEB Mean |
|-------|--------|------|-----------|
| **Octen-Embedding-8B** | 7.6B | 4096 | **81.06** |
| Octen-Embedding-4B | 4.0B | 2560 | 78.75 |
| Qwen3-Embedding-8B | 7.6B | 4096 | 73.73 |
| inf-retriever-v1-1.5b | 1.5B | 1536 | ~45 (5 tasks) |

The MacBook M4 Max (128GB unified memory, 546GB/s bandwidth) can comfortably run the 8B model at 14.4GB memory footprint.

## Proposal

1. Run A/B comparison: `inf-retriever-v1-1.5b` vs `Octen-Embedding-8B` on NEXUS's existing eval corpus (~800-900k words frozen story)
2. Measure retrieval accuracy on representative queries
3. Measure latency impact (embedding generation + search)
4. If results justify it, upgrade the pipeline

pgvector 0.6.0 now supports up to 16,000 dimensions, so the 4096-dim embeddings are no longer an issue.

## Notes
- This is a potential upgrade path identified during EDI memory system design
- Rigorous A/B validation required before any changes
- Consider also testing Octen-4B as a middle ground (2560 dims, 78.75 RTEB)

**Comment by pythagorakase** [comment]
## Decision Needed: HNSW Index Strategy for Octen Tables

The IR eval v2 framework (PR #179, #180) is now merged and ready to run head-to-head comparisons. Before octen embeddings can be generated and evaluated, we need to resolve a pgvector dimension-cap question that affects both candidate models.

### Background

Issue #175's candidate models are:
- **Octen-Embedding-4B** — 2,560 dimensions
- **Octen-Embedding-8B** — 4,096 dimensions

`migrations/014_add_2560d_4096d_embeddings.sql` creates the destination tables and an HNSW index on each:

```sql
CREATE TABLE chunk_embeddings_2560d (..., embedding vector(2560), ...);
CREATE TABLE chunk_embeddings_4096d (..., embedding vector(4096), ...);

CREATE INDEX ... ON chunk_embeddings_2560d USING hnsw (embedding vector_cosine_ops);
CREATE INDEX ... ON chunk_embeddings_4096d USING hnsw (embedding vector_cosine_ops);
```

This migration has not been applied to any database yet (verified across ` NEXUS`, ` NEXUS_template`, and all save slots — latest applied migration is 012).

### The Blocker

We're on pgvector 0.8.0. Per pgvector's documented limits:

| Column type | Storage cap | HNSW indexable cap |
|---|---|---|
| `vector` | 16,000 | **2,000** |
| `halfvec` (16-bit float) | 16,000 | **4,000** |
| `bit` (binary) | 64,000 | 64,000 |
| `sparsevec` | 1B nz | 1,000 nz |

Both octen tables exceed the 2,000-dim cap on `vector` HNSW. The 4,096-dim table also exceeds halfvec's 4,000-dim cap (by 96 dims).

**If the docs reflect reality on 0.8.0, applying migration 014 will fail with**: `ERROR: column cannot have more than 2000 dimensions for hnsw index`.

### Quick Primer (for reviewers unfamiliar with pgvector internals)

- **HNSW** (Hierarchical Navigable Small World): approximate-nearest-neighbor index. Sub-millisecond retrieval. The dim cap comes from the internal page format — vectors are stored inline in index pages, which have a bounded size.
- **halfvec**: a pgvector column type that uses 16-bit half-precision floats instead of 32-bit single-precision. Halves the per-vector size, which is why the indexable cap doubles (4,000 vs 2,000).
- **Sequential scan**: no index. Each query computes distance to every stored vector. Linear in corpus size, no dim cap.
- **Why corpus size matters here**: a typical NEXUS save slot has on the order of 1k–10k narrative chunks. At 4,096 dims, sequential scan over 1k vectors is roughly 5–15 ms per query. HNSW would be sub-millisecond. The difference is real but likely imperceptible to a human running an eval over ~50 queries.

### Options

| # | Strategy | Octen-4B (2560d) | Octen-8B (4096d) | Pros | Cons |
|---|---|---|---|---|---|
| 1 | **Empirically test pgvector caps first** | TBD | TBD | Ground truth — if the cap was lifted in 0.8.0, the migration just works as-is. Cheap to verify (a `CREATE INDEX` on a tiny test table). | Adds a verification step before any other work. |
| 2 | **Drop all HNSW indexes** | `vector(2560)`, seq-scan | `vector(4096)`, seq-scan | Works regardless of pgvector version. Simplest fix. No type-system changes. | ~10ms latency per query vs sub-ms with HNSW. Probably invisible at eval scale. |
| 3 | **Mixed: halfvec for 4B, no-index for 8B** | `halfvec(2560)` HNSW | `vector(4096)`, seq-scan | Octen-4B gets fast indexed retrieval; both models still testable | Asymmetric performance comparison; `EmbeddingManager` may need a type adapter for `halfvec` columns. |
| 4 | **Scope down to Octen-4B only** | `halfvec(2560)` HNSW | (skipped) | Smallest surface area; clean indexed retrieval. | Loses head-to-head data on the larger and more interesting model — defeats half the purpose of #175. |

### My Read

For an *evaluation* (where we run ~50 queries by a human or LLM judge, not a production hot path), option 2 (drop indexes) seems like the right call: simplest, avoids type-system entanglement, and a 10 ms vs sub-ms gap is invisible at eval scale. If octens win and we promote to production, *that's* the time to revisit indexing — by then we'd also know exactly which dim count we're committing to.

But option 1 (empirical test) is so cheap it's worth running first regardless. If pgvector 0.8.0 has actually lifted the `vector` HNSW cap, the whole decision collapses.

### Asks for Reviewers

1. **Is my reading of pgvector 0.8.0's HNSW caps correct?** If you know of a recent change that raised them on `vector` (not just `halfvec`), please flag — that resolves this immediately.
2. **Does halfvec's reduced precision (fp16 vs fp32) materially affect retrieval quality** for embedding similarity at this scale? My intuition is "no, the noise floor is below model-to-model variance," but I haven't tested.
3. **Any reason to keep an HNSW index during eval at all**, given corpus size? E.g., does the IR eval framework run enough queries that 10 ms per query becomes annoying?

**Comment by pythagorakase** [comment]
## Empirical verification (per reviewer suggestion)

Ran option 1 against the dev install (pgvector 0.8.0). Caps match the docs exactly — no surprises:

| Test | Result |
|---|---|
| `vector(2560)` HNSW | FAIL: `column cannot have more than 2000 dimensions for hnsw index` |
| `vector(4096)` HNSW | FAIL: same error |
| `halfvec(2560)` HNSW | OK |
| `halfvec(4096)` HNSW | FAIL: `column cannot have more than 4000 dimensions for hnsw index` |
| `vector(2000)` (cap boundary) | OK |
| `vector(2001)` (cap boundary) | FAIL |
| `halfvec(4000)` (cap boundary) | OK |
| `halfvec(4001)` (cap boundary) | FAIL |

So both Octen tables in migration 014 exceed the HNSW caps as written, and even rewriting to halfvec wouldn't save the 4096-dim case (off by 96 dims).

**Proceeding with option 2** (drop HNSW indexes for both new tables, sequential scan only) per consensus. At eval scale (~1k–10k chunks, ~50 queries per run) the latency cost is on the order of 10 ms/query — invisible against the cost of LLM-judged relevance. If octens win and we promote to production, the index strategy is the right thing to revisit then, with concrete corpus and query-volume numbers in hand.

**Comment by pythagorakase** [comment]
## Bakeoff complete — Octen-Embedding-4B wins (the 8B does not)

Ran all three fp Octen variants (0.6B / 4B / 8B) head-to-head against the strongest current baseline (`inf-retriever-v1-1.5b`, run 6). Full IR eval framework, 130 golden queries, judge-on-demand turned on so all 1300 query-chunk pairs per run got real relevance scores.

### Quality (vs. baseline run 6, sign convention: positive = octen better)

| Metric | Baseline 1.5B | Octen-0.6B Δ | Octen-4B Δ | Octen-8B Δ |
|---|---|---|---|---|
| P@5 | 0.886 | −0.012 | **+0.015** | +0.009 |
| P@10 | 0.845 | −0.006 | **+0.015** | +0.012 |
| MRR | 0.952 | +0.012 | **+0.025** | +0.020 |
| bpref | 0.222 | −0.002 | **+0.003** | +0.000 |
| NDCG@10 | 0.703 | −0.015 | **+0.021** | +0.014 |

**Octen-4B sweeps every metric directionally.** None survive p<0.05 at n=130 (smallest p-values 0.14 on MRR and NDCG), but the direction-consistency across 5 independent metrics rules against noise. 0.6B is statistically tied with baseline (efficiency play). 8B is *worse* than 4B on every metric — sweet-spot inversion, likely from over-parameterization on a 1,425-chunk corpus.

### Latency (separate microbenchmark, one query → top-10, production-like path)

| Config | Median total | Median encode | Median SQL | Median rerank |
|---|---|---|---|---|
| Current ensemble (bge + e5 + inf-retriever) | 9,428 ms | 184 ms | 364 ms | 8,869 ms |
| **Octen-4B solo** | **6,498 ms** | 210 ms | 365 ms | 5,827 ms |
| inf-retriever-1.5b solo | 5,419 ms | 59 ms | 357 ms | 4,981 ms |

**Swapping to Octen-4B solo cuts user-perceived latency by ~31%** (−2.9s/query) with comparable-or-better retrieval quality.

### Decision pointer

When ready to swap, the move is:
1. Set `is_active = false` on `bge-large-en`, `e5-large-v2`, `inf-retriever-v1-1.5b` in `[memnon.models.*]`.
2. Set `is_active = true`, `weight = 1.0` on `[memnon.models."Octen-Embedding-4B"]`.
3. Optional: keep ensemble plumbing in place in case we ever want to add 4B as a member alongside the others rather than as a replacement.

### Related follow-up

The latency benchmark surfaced that **cross-encoder reranking is 85-95% of total query latency** across all configs. Branch `feature/expand-retrieval-benchmarks` adds infrastructure to run a similar bakeoff on the reranker side: Qwen3-Reranker-0.6B / Qwen3-Reranker-4B / mxbai-rerank-large vs. the current DeBERTa-v3 (which is from TREC-DL 2022, same era as the embeddings we're replacing).

### Skipped

INT8 quantized variants (`Octen-Embedding-4B-INT8`, `Octen-Embedding-8B-INT8`) were skipped — bitsandbytes has no MPS backend on Apple Silicon, and CPU-only inference would be ~10× slower than the fp variants. Registered in `nexus.toml` for completeness if anyone needs them later.

**Comment by pythagorakase** [comment]
## Reranker bakeoff (follow-on): Qwen3-Reranker-4B wins on quality, but at a latency cost

After concluding the embedder bakeoff (Octen-Embedding-4B winning above), I noticed the cross-encoder reranker (`naver-trecdl22-crossencoder-debertav3`) is from 2022 — same vintage as our prior 2025 embedding picks. Since SOTA reranker architectures moved decisively in 2024-2026 (LLM-backed yes/no rerankers like Qwen3-Reranker), ran the same bakeoff pattern on the reranker side.

### Method

Same 130 golden queries, same Octen-Embedding-4B retriever, same `--top-k 10` retrieval+rerank shape. Only the reranker varies. Compared each to **run 10 (`octen-4b-solo`)** which uses the production DeBERTa-v3 reranker.

### Quality (delta favors A = DeBERTa-v3 better)

| Reranker | P@5 Δ | P@10 Δ | MRR Δ | bpref Δ | NDCG@10 Δ | Verdict |
|---|---|---|---|---|---|---|
| **Qwen3-Reranker-0.6B** (run 15) | -0.008 | +0.003 | -0.0003 | **+0.004 (sig)** | +0.007 | Statistically tied |
| **mxbai-rerank-large-v2** (run 16) | **+0.046 (sig)** | +0.001 | **+0.069 (sig)** | **+0.009 (sig)** | **+0.050 (sig)** | **Worse than DeBERTa-v3** |
| **Qwen3-Reranker-4B** (run 17) | **-0.022 (sig)** | -0.002 | -0.005 | +0.001 | **-0.008 (sig)** | **Better than DeBERTa-v3** |

- Negative delta = candidate (B) wins
- Significance: paired permutation test, p<0.05

### Wall-time (CPU/MPS effects matter here)

| Reranker | Run wall time | Hardware notes |
|---|---|---|
| DeBERTa-v3 (production) | ~10 min | MPS (fast, standard CrossEncoder) |
| Qwen3-Reranker-0.6B | 15 min | **CPU** (Qwen3 hits an MPS Metal bug; see below) |
| mxbai-rerank-large-v2 | 7 min | MPS (standard CrossEncoder, fastest) |
| Qwen3-Reranker-4B | 75 min | **CPU** (same MPS issue, more parameters) |

### The MPS issue

Qwen3-Reranker inference on real-length narrative chunks triggers a Metal assertion:
```
MPSNDArray.mm: NDArray dimension length > INT_MAX
```
Reproduces with both SDPA and eager attention, fp16 and bf16, max_length 2048 or 8192. Smoke tests with short toy passages mask it — real NEXUS chunks (200-6300 chars) hit it consistently. **Workaround: CPU inference.** That's the 7-15x latency penalty visible above.

### Decision: keep DeBERTa-v3 reranker in production, despite Qwen3-4B's quality win

Reasoning:
- The quality lift from Qwen3-4B is real (P@5 +0.022, NDCG +0.008, both p<0.05) but small
- The latency cost on Apple Silicon is huge (~7x slower on CPU)
- The Octen-4B embedder swap already cut overall latency by 31%; flipping the reranker back to a 7x-slower path would erase that win and then some
- Production runs on Apple Silicon today; CUDA would change the calculus entirely

**The infrastructure is in place to swap whenever:**
- PyTorch fixes the MPS Qwen3 bug, OR
- We move to CUDA hosting

The `[cross_encoder_reranking.candidates.*]` registry in `nexus.toml` and the `--reranker NAME` flag on `ir_eval.runner create-run` let anyone re-run the bakeoff cheaply once the environment changes.

### Surprise findings worth noting

1. **mxbai-rerank-large-v2 underperformed badly** — it's a 2025-era reputable reranker, but trained on standard IR benchmarks (BEIR-style web search). Our narrative-fiction domain is far enough from that distribution that the older DeBERTa-v3 transfers better. A reminder that benchmark leaderboards aren't always a proxy for your use case.

2. **Qwen3-Reranker-0.6B matched DeBERTa-v3-large at ~7x fewer parameters.** The LLM yes/no paradigm is competitive with traditional cross-encoders at small scales. If we ever need a tiny reranker (mobile, low-resource), 0.6B is a real option once MPS support stabilises.

3. **The eval framework reranks 10 candidates, not 30 like production does.** Both runs use the same shape so the comparison is fair, but the gap might widen with a deeper candidate pool. Worth re-running on a CUDA box with `--top-k 30` if revisiting.

### What landed on `feature/expand-retrieval-benchmarks`

- `0eb3864` Octen embedding plumbing (5 variants, migration 014 HNSW removal, INT8 CPU dispatch)
- `4bd12cc` Reranker candidate registry + `Qwen3LMReranker` class + dispatch
- `d436cd9` Embedder swap to Octen-Embedding-4B (the #175 conclusion)
- `a7ff688` Qwen3 MPS workaround (CPU + max_length=2048 + bf16 default)
